new-tutorial/2_dglgraph.py

```diff
@@ -0,0 +1,211 @@
+# .. note::
+# 
+#    The graphs in DGL are always directed since messages sent
+#    from one node to the other are often different between both
+#    directions. If you want to handle undirected graphs, you may consider
+#    treating it as a bidirectional graph. See `Graph
+#    Transformations`_ for an example of making
+#    a bidirectional graph.
+# 
+
+
+######################################################################
+# Assigning Node and Edge Features to Graph
+# -----------------------------------------
+# 
+# In many graph data, nodes and edges have attributes. The nodes and edges
+# in DGL graphs can have multiple attributes.
+# 
+# Although the types of node and edge attributes can be arbitrary in real
+# world, a DGL graph only accepts attributes stored in tensors (with
+# numerical contents). Consequently, an attribute of all the nodes or
+# edges must have the same shape. In the context of deep learning, those
+# attributes are often called *features*.
+# 
+# You can assign and retrieve node and edge features via ``ndata`` and
+# ``edata`` property.
+# 
+
+# Assign a 3-dimensional node feature vector for each node.
+g.ndata['x'] = torch.randn(6, 3)
+# Assign a 4-dimensional edge feature vector for each edge.
+g.edata['a'] = torch.randn(5, 4)
+# Assign a 5x4 node feature matrix for each node.  Node and edge features in DGL can be multi-dimensional.
+g.ndata['y'] = torch.randn(6, 5, 4)
+
+print(g.edata['a'])
+
+
+######################################################################
+# .. note::
+# 
+#    The vast development of deep learning has provided us many
+#    ways to encode various types of attributes into numerical features.
+#    Here are some general suggestions:
+# 
+#    -  For categorical attributes (e.g. gender, occupation), consider
+#       converting them to integers or one-hot encoding.
+#    -  For variable length string contents (e.g. news article, quote),
+#       consider applying a language model.
+#    -  For images, consider applying a vision model such as CNNs.
+# 
+#    You can find plenty of materials on how to encode such attributes
+#    into a tensor in the `PyTorch Deep Learning
+#    Tutorials <https://pytorch.org/tutorials/>`__.
+# 
+
+
+######################################################################
+# Graph Queries
+# -------------
+# 
+# ``DGLGraph`` object provides various methods to query a graph structure.
+# 
+
+print(g.num_nodes())
+print(g.num_edges())
+# Out degrees of the internal node
+print(g.out_degrees(0))
+# In degrees of the internal node - note that the graph is directed so the in degree should be 0.
+print(g.in_degrees(0))
+
+
+######################################################################
+# Graph Transformations
+# ---------------------
+# 
+
+
+######################################################################
+# You can also transform a graph to another graph.
+# 
+# A common transformation is to create a subgraph from the original graph:
```

Comment:
DGL provides many APIs to transform a graph to another such as extracting a subgraph from the original one: